datahub/dbmaintenance/management/commands/update_one_list_fields.py

```
@@ -0,0 +1,164 @@
+        return Advisor.objects.get(pk=pk)
+
+    def _update_company(
+        self, company, new_classification_id, new_one_list_account_owner_id, simulate
+    ):
+        """
+        Update `company` with the new values.
+
+        :param new_classification_id: new CompanyClassification value
+        :param new_one_list_account_owner_id: new Advisor value
+        :param simulate: True if the change should not be committed to the database
+        """
+        company.classification = self.get_classification(new_classification_id)
+        company.one_list_account_owner = self.get_adviser(new_one_list_account_owner_id)
+
+        if simulate:
+            return
+
+        with reversion.create_revision():
+            company.save(
+                update_fields=(
+                    'classification_id',
+                    'one_list_account_owner_id',
+                )
+            )
+            reversion.set_comment('Classification and One List account owner correction.')
+
+    def _should_update(self, company, classification_id, one_list_account_owner_id):
+        """
+        Check if `company` should be updated.
+
+        :param company: Company to update
+        :param classification_id: new CompanyClassification value
+        :param one_list_account_owner_id: new Advisor value
+        :return: True if `company` needs updating
+        """
+        return (
+            company.classification_id != classification_id
+        ) or (
+            company.one_list_account_owner_id != one_list_account_owner_id
+        )
+
+    def _process_row(self, row, simulate=False, **options):
+        """Process one single row."""
+        company = Company.objects.get(pk=row['id'])
+        classification_id = parse_uuid(row['classification_id'])
+        one_list_account_owner_id = parse_uuid(row['one_list_account_owner_id'])
+
+        # remove company.pk from the list of companies to reset
+        self.companies_to_reset.pop(company.pk, None)
```

If there's an error earlier in the function, how does company get removed from `self.companies_to_reset`?

---

What kind of error? I could move this to the line immediately after getting the company maybe?

---

Any error, doesn't CSVBaseCommand.process_row() swallow up exceptions?

I'm not sure the ignore errors behaviour makes sense with --reset-unmatched.

---

I'm not sure to understand:

- `handle` calls `process_row` which wraps `_process_row` in a try/except
- we want to avoid resetting the company, even if there are exceptions so we still want to exclude that company from the `self.companies_to_reset` dict before any logic.

I wanted to wrap everything into an atomic transaction but keeping the same logic that we have always based on is nice.

I actually moved that line a bit further up which looks a lot better.

---

It's better with `self.companies_to_reset.pop(company.pk, None)` at the top but it's still weird to have that in an exception handler. I'll click approve though since there's less risk having it at the top.